tic-tac-toe/src/urbit/app/toe.hoon

```diff
@@ -0,0 +1,1152 @@
+::  Tic-Tac-Toe
+::
+/-  *toe, sole
+/+  sole, *server, *toe
+::
+:: This imports the tile's JS file from the file system as a variable.
+::
+/=  tile-js
+  /^  octs
+  /;  as-octs:mimes:html
+  /:  /===/app/toe/js/tile
+  /|  /js/
+      /~  ~
+  ==
+::
+=,  format
+=,  sole
+::
+=>  |%
+    ::
+    +|  %models
+    ::
+    +$  state
+      $:  ::  (see %/sur/toe/hoon)
+          ::
+          toers=players
+          subs=subscribers
+          game=game-state
+          board=board-state
+          ::  $who: player that can perform a move
+          ::
+          who=ship
+          ::  $next: flag to indicate a replay
+          ::
+          next=?
+          ::
+          ::  $consol: console state
+          ::
+          ::     $conn:  id for console connection
+          ::     $state: data in the console
+          ::
+          consol=[conn=bone state=sole-share]
+      ==
+    ::
+    +$  move  (pair bone card)
+    ::
+    +$  card
+      $%  [%diff diff-data]
+          [%peer wire dock path]
+          [%pull wire dock ~]
+          [%poke wire dock poke-data]
+          [%wait wire p=@da]
+          [%http-response =http-event:http]
+          [%connect wire binding:eyre term]
+      ==
+    ::
+    +$  diff-data
+      $%  ::  See %/mar/toe/* for specific details
+          ::    * = [turno, player, winner]
+          ::
+          [%toe-turno toe-turno]
+          [%toe-player toe-player]
+          [%toe-winner toe-winner]
+          [%sole-effect sole-effect]
+          [%json json]
+       ==
+    ::
+    +$  poke-data
+      $%  [%toe-cancel toe-cancel]
+          [%launch-action [@tas path @t]]
+      ==
+    ::
+    ::  FIXME: $spot has to be redefined
+    ::     even though it's already in sur...
+    ::     it gave nest fails in +-moves-start
+    ::     ?:  (~(has by board.sat) (spot-val u.try))
+    ::     -find.get
+    ::
+    +$  spot  [coord coord]
+    --
+::
+::  %app
+::
+::    Our app is defined as a "door" (multi-armed core with a sample).
+::    Arms are grouped in chapters (+|) based on common functionality
+::
+|_  [bol=bowl:gall %0 sat=state]
+::
+::  %alias: rewording of commonly used nouns
+::
++|  %alias
+::
+::  %me: alias for our ship
+::
+++  me  our.bol
+::
+::  %ze: alias for the origin of an incoming request
+::
+++  ze  src.bol
+::
+::
+::  %this: common idiom to refer to our whole %app door and its context
+::
+++  this  .
+::
+::  Aliasing arms declared within cores (requiered by gall)
+::
+++  poke-sole-action          poke-sole-action:co:view
+++  peer-sole-action          peer-sole-action:co:view
+++  peer-sole                 peer-sole:co:view
+++  peer-toetile              peer-toetile:fe:view
+++  poke-json                 poke-json:fe:view
+++  bound                     bound:fe:view
+++  poke-handle-http-request  poke-handle-http-request:fe:view
```

Comment:
I feel that this is very ugly... (Furthermore, it's close to `++  this  .` which is frowned upon in HoonSchool)

Reply:
This isn't too bad. If there's any logic in those arms that deal exclusively with gall-side aspects (wire unpacking, etc) then probably move that here, but I think this is fine as-is.

> `++ this .` which is frowned upon in HoonSchool)

But that's just because @baudtack is a grump. (; The alternative is `+>.$` or `..some-arm` all over the place, which isn't much better. (You might want to use `+*` as special notation for these "alias" arms, but [it's complicated](https://github.com/urbit/urbit/pull/1654#issuecomment-527637112).)

Reply:
That discussion came up in Hoon201 :) I also tried with `+*` but it seems that it can't be used within doors.

Initially I used `+>.$`, then looked at some of the newer apps, saw `this` used there and replaced it, thinking "I'm following best practices!". So wrong!